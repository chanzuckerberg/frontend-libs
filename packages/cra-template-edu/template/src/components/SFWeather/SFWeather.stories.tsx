import type {StoryObj} from '@storybook/react';
import {within} from '@storybook/testing-library';
import {rest} from 'msw';
import WithMockRequests from '../../testing/WithMockRequests';
import SFWeather from './SFWeather';

export default {
  component: SFWeather,
};

const handlers = [
  rest.get('https://api.open-meteo.com/v1/forecast', (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({current_weather: {temperature: 66.6}}),
    );
  }),
];

export const Standard: StoryObj = {
  render: () => (
    <WithMockRequests requestHandlers={handlers}>
      <SFWeather />
    </WithMockRequests>
  ),
  play: async ({canvasElement}) => {
    const canvas = within(canvasElement);
    // Wait for the API request to finish before doing anything else (such as a11y testing, visual
    // regression testing, etc).
    await canvas.findByText('Current San Francisco temp: 66.6');
  },
};
