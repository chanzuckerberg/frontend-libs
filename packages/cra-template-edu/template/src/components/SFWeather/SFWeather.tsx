import {useQuery} from '@tanstack/react-query';
import {getTemp} from '../../api/weather';

const sfCoord = {lat: 37.77, long: -122.43};

export default function SFWeather() {
  const query = useQuery(
    // Query key, which should be unique to this data. Include any query params or ids.
    // https://tanstack.com/query/v4/docs/guides/query-keys
    ['temp', sfCoord.lat, sfCoord.long],
    // Function that actually makes the HTTP request.
    // https://tanstack.com/query/v4/docs/guides/query-functions
    () => getTemp(sfCoord.lat, sfCoord.long),
  );

  return <p>Current San Francisco temp: {query.data ?? 'ðŸ¤”'}</p>;
}
