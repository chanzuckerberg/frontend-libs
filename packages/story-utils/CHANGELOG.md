# Changelog

## Unreleased

- [breaking] Output ES2019 instead of ES6 - [#19](https://github.com/chanzuckerberg/frontend-libs/pull/19)
- [breaking] Output [ES modules](https://nodejs.org/api/esm.html) instead of CommonJS - [#19](https://github.com/chanzuckerberg/frontend-libs/pull/19)
- [breaking] The `argOverrides` option for `generateSnapshot` has been removed
- [breaking] Added new peer dependency on any version of React. This was already implicit (since React testing-library is being used)
- [breaking] `prepareStory` has been removed. @storybook/testing-react can be used instead
- [breaking] `getStories` has been removed, to simplify supporting Storybook 6.4. It's not used much
- [breaking] Rename the `getSnapshot` option of `generateSnapshot` to `getElement`, to better describe what it is for

## 2.1.0 (2021-08-13)
- Update dependencies
- [new] Add `wait` helper - [#17](https://github.com/chanzuckerberg/frontend-libs/pull/17)
- [fix] Push generating snapshots to the end of the event loop in `generateSnapshots` - [#17](https://github.com/chanzuckerberg/frontend-libs/pull/17)

## 2.0.0 (2021-03-04)

- [breaking] Change the `generationSnapshot` option `disabled` to `skip` - [#13](https://github.com/chanzuckerberg/frontend-libs/pull/13)
- [maintenance] Upgrade dependencies - [#14](https://github.com/chanzuckerberg/frontend-libs/pull/14)

## 1.2.1 (2021-02-16)

- [maintenance] Run `update-ts-references` on CI for Windows - [#11](https://github.com/chanzuckerberg/frontend-libs/pull/11)

## 1.2.0 (2020-12-16)

- [new] Added `generateSnapshots` helper - [#8](https://github.com/chanzuckerberg/frontend-libs/pull/8)

## 1.1.0 (2020-12-04)

- [new] Validate story metadata when finding stories with `getStories`

## 1.0.0 (2020-12-03)

Initial release! ðŸŽ‰
