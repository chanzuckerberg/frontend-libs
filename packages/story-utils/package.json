{
  "name": "@chanzuckerberg/story-utils",
  "version": "2.0.0",
  "license": "MIT",
  "keywords": [
    "stories",
    "storybook",
    "testing"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/chanzuckerberg/frontend-libs"
  },
  "homepage": "https://github.com/chanzuckerberg/frontend-libs/blob/main/packages/story-utils",
  "publishConfig": {
    "access": "public"
  },
  "main": "build/index.js",
  "files": [
    "build"
  ],
  "peerDependencies": {
    "@storybook/react": "^6.0.0",
    "@testing-library/react": "^11.0.0"
  },
  "dependencies": {
    "glob": "^7.1.7",
    "lodash": "^4.17.21",
    "zod": "^1.11.17"
  }
}
