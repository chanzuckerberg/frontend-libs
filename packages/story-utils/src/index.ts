export { default as getStories } from './getStories';
export { default as prepareStory } from './prepareStory';
