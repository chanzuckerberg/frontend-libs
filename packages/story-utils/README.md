# story-utils

Utilities for using [Storybook stories](https://storybook.js.org/docs/react/get-started/whats-a-story) in non-Storybook environments. For example, using stories in Jest tests.

## Installation

Install @chanzuckerberg/story-utils with your favorite package manager. With yarn, that would look like

```sh
yarn add --dev @chanzuckerberg/story-utils
```

## Utilities

### getStories

Get all stories from the local filesystem. Useful for consuming stories in some way in a Node environment.

```js
import { getStories } from '@chanzuckerberg/story-utils';

const stories = getStories('src/components/**/*.stories.tsx');

const storyNames = stories.map((story) => story.name);
console.log('Found stories:', storyNames);
```

### prepareStory

Take a story and return its React elements for unit testing.

For example, imagine you have this story

```js
// MyButton.stories.jsx

export const MyStory = (...args) => <Button {...args} />;

MyStory.args = {
  children: 'button!',
};
```

This utility makes it easy to re-use the story in tests.

```js
// MyButton.test.jsx

// Using testing-library to demonstrate.
import { render, screen } from '@testing-library/react';
import { prepareStory } from '@chanzuckerberg/story-utils';
import { MyStory } from './MyButton.stories';

test('my button', () => {
  render(prepareStory(MyStory));
  expect(screen.getByText('button!')).toBeTruthy();
});
```

Overrides for specific args can also be provided, to customize a test.

```js
test('my button 2', () => {
  // Can pass overrides
  render(prepareStory(MyStory, {
    children: 'coffee!',
  }));
  expect(screen.getByText('coffee!')).toBeTruthy();
});
```
