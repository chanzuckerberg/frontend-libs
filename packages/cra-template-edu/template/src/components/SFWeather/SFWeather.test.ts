/**
 * @jest-environment jsdom
 */

import {generateSnapshots} from '@chanzuckerberg/story-utils';
import * as stories from './SFWeather.stories';

generateSnapshots(stories);
