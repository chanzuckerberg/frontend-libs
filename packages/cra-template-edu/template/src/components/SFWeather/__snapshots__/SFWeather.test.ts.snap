// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Standard story renders snapshot 1`] = `
<p>
  Current San Francisco temp: 
  66.6
</p>
`;
